src/porepy/models/applications/terzaghi_model.py

```diff
@@ -0,0 +1,706 @@
+        data[pp.STATE][self.scalar_variable] = initial_p
+        data[pp.STATE][pp.ITERATE][self.scalar_variable] = initial_p
+        self.solutions.append(TerzaghiSolution(self))
+
+    def _bc_type_scalar(self, sd: pp.Grid) -> pp.BoundaryCondition:
+        """Define boundary condition types for the flow subproblem.
+
+        Args:
+            sd: Subdomain grid.
+
+        Returns:
+            Scalar boundary condition representation.
+
+        """
+        # Define boundary regions
+        all_bc, _, _, north, *_ = self._domain_boundary_sides(sd)
+        north_bc = np.isin(all_bc, np.where(north)).nonzero()
+
+        # All sides Neumann, except the North which is Dirichlet
+        bc_type = np.asarray(all_bc.size * ["neu"])
+        bc_type[north_bc] = "dir"
+
+        bc = pp.BoundaryCondition(sd, faces=all_bc, cond=list(bc_type))
+
+        return bc
+
+    def _bc_type_mechanics(self, sd: pp.Grid) -> pp.BoundaryConditionVectorial:
+        """Define boundary condition types for the mechanics subproblem
+
+        Args:
+            sd: Subdomain grid.
+
+        Returns:
+            Vectorial boundary condition representation.
+
+        """
+        # Inherit bc from parent class. This sets all bc faces as Dirichlet.
+        super()._bc_type_mechanics(sd=sd)
+
+        # Get boundary sides, retrieve data dict, and bc object
+        _, east, west, north, south, *_ = self._domain_boundary_sides(sd)
+        data = self.mdg.subdomain_data(sd)
+        bc = data[pp.PARAMETERS][self.mechanics_parameter_key]["bc"]
+
+        # East side: Roller
+        bc.is_neu[1, east] = True
+        bc.is_dir[1, east] = False
+
+        # West side: Roller
+        bc.is_neu[1, west] = True
+        bc.is_dir[1, west] = False
+
+        # North side: Neumann
+        bc.is_neu[:, north] = True
+        bc.is_dir[:, north] = False
+
+        # South side: Roller
+        bc.is_neu[0, south] = True
+        bc.is_dir[0, south] = False
+
+        return bc
+
+    def _bc_values_mechanics(self, sd: pp.Grid) -> np.ndarray:
+        """Set boundary condition values for the mechanics subproblem.
+
+        Args:
+            sd: Subdomain grid.
+
+        Returns:
+            Vectorial boundary condition values of shape (sd.dim * sd.num_faces, ).
+
+        """
+
+        # Retrieve boundary sides
+        _, _, _, north, *_ = self._domain_boundary_sides(sd)
+
+        # All zeros except vertical component of the north side
+        vertical_load = self.params["vertical_load"]
+        bc_values = np.array([np.zeros(sd.num_faces), np.zeros(sd.num_faces)])
+        bc_values[1, north] = -vertical_load * sd.face_areas[north]
+        bc_values = bc_values.ravel("F")
+
+        return bc_values
+
+    def after_newton_convergence(
+        self,
+        solution: np.ndarray,
+        errors: float,
+        iteration_counter: int,
+    ) -> None:
+        """Method to be called after the linear system has been solved."""
```

Comment:
nonlinear system, right?

Reply:
Yes. Although that would be somewhat misleading since the problem is linear. I will just avoid such details.